### ISSUE TITLE
Casting embeds with empty map does not validate data already in the struct

### ISSUE THREAD

**Issue by mracos** [elixir-ecto/ecto#3854]
### Environment

* Elixir version (elixir -v): `1.13.0`
* Database and version (PostgreSQL 9.4, MongoDB 3.2, etc.): N/A
* Ecto version (mix deps):  `~> 3.5`
* Database adapter and version (mix deps): N/A
* mix.lock: "ecto": `{:hex, :ecto, "3.7.2", "44c034f88e1980754983cc4400585970b4206841f6f3780967a65a9150ef09a8"}`
* Operating system: Mac OS Monterey 12.1 

### Current behavior

When casting a regular struct with an empty map in the `params` it validates data that is not in changes and was previously set in the struct, however, this does not happen for embedded schemas unless we add an empty nested map to the top-level params.

e.g. with these two modules

```elixir
defmodule SongMetadata do
  use Ecto.Schema
  import Ecto.Changeset

  @primary_key false
  embedded_schema do
    field(:released, :date)
  end

  def changeset(song_metadata \\ %__MODULE__{}, params) do
    song_metadata
    |> cast(params, [:released])
    |> validate_required([:released])
  end
end
```
```elixir
defmodule Song do
  use Ecto.Schema
  import Ecto.Changeset

  schema "songs" do
    field(:name, :string)
    embeds_one(:metadata, SongMetadata, on_replace: :update)
  end

  def changeset(song \\ %__MODULE__{}, params) do
    song
    |> cast(params, [:name])
    |> cast_embed(:metadata)
    |> validate_required([:name])
  end
end
```

If one casts `Song` with an empty map, it validates that `:name` is required **even if it was  not included in the params**, however, we can see that no validation was made for the `embedded`:

```elixir
iex(1)> Song.changeset(%Song{}, %{})
#Ecto.Changeset<
  action: nil,
  changes: %{},
  errors: [name: {"can't be blank", [validation: :required]}],
  data: #Song<>,
  valid?: false
>

iex(2)> Song.changeset(%Song{name: "asas"}, %{})
#Ecto.Changeset<action: nil, changes: %{}, errors: [], data: #Song<>, valid?: true>
```

However if one adds an empty map for the embedded, it does the validations, (although it also appears in the `changes` field for the changeset, not sure if that was supposed to happen).

```elixir
iex(3)> Song.changeset(%Song{}, %{"metadata" => %{}})
#Ecto.Changeset<
  action: nil,
  changes: %{
    metadata: #Ecto.Changeset<
      action: :insert,
      changes: %{},
      errors: [released: {"can't be blank", [validation: :required]}],
      data: #SongMetadata<>,
      valid?: false
    >
  },
  errors: [name: {"can't be blank", [validation: :required]}],
  data: #Song<>,
  valid?: false
>
```

### Expected behavior

Validate the embedded schema even if not passing an empty nested map with the same name as the embedded params.

The use case here is to validate that some data fetched from the database (thus already included in the struct) is valid even if there are no new changes, if I'm approaching the problem wrong and changesets were not supposed to be used this way please let me know! 🥺

**Comment by mracos** [comment]
But I also just saw that some validations are not applied to data already in the struct (maybe that is specific to `validate_required`)?

e.g. with the same modules above but adding a `|> validate_length(:name, min: 2)` to `Song.changeset/2`

```elixir
iex(4)> Song.changeset(%Song{name: "1"}, %{})
#Ecto.Changeset<action: nil, changes: %{}, errors: [], data: #Song<>, valid?: true>

iex(5)> Song.changeset(%Song{name: "1"}, %{name: "2"})
#Ecto.Changeset<
  action: nil,
  changes: %{name: "2"},
  errors: [
    name: {"should be at least %{count} character(s)",
     [count: 2, validation: :length, kind: :min, type: :string]}
  ],
  data: #Song<>,
  valid?: false
>
```

**Comment by josevalim** [comment]
This is correct. cast_embed is only going to cast if a map is given. You either need to force the map or pass `required: true` to cast_embed.

cast is also always about validating external data. It won't validate what is already there. :)

**Comment by mracos** [comment]
I see, tyvm for the answer 💙
Forgot about the `required: true` 🤦

**Comment by mracos** [comment]
But how should I approach it if I want to be able to validate what is already there? I derive a `map` from the data to pass as changes? Or I am using `changesets` in a way that was not designed to be used? 🤔

**Comment by josevalim** [comment]
Change sets are about validating external data. If you have invalid data currently in your struct, then you need to ask yourself who put that invalidated there in the first place because you're not supposed to have invalidating your structs.

**Comment by josevalim** [comment]
Exactly. The changesets assume the struct is valid.
